src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java

```diff
@@ -0,0 +1,382 @@
+                        choosing.start();
+                        for (GATKRead r : rds) {
+                            if (!accept(r, i)) {
+                                rejected++;
+                                continue;
+                            }
+                            while (currentShard.getEnd() < r.getStart()) {
+                                // move to the next shard.
+                                if (!ret.isEmpty()) {
+                                    accepted += ret.size();
+                                    choosing.stop();
+                                    shipping.start();
+                                    kving.start();
+                                    KV<String, Iterable<GATKRead>> out = KV.of(currentShard.toString(), ret);
+                                    kving.stop();
+                                    outputting.start();
+                                    c.output(out);
+                                    outputting.stop();
+                                    //bunny.stepEnd("output subshard " + currentShardIndex + ", " + ret.size() + " reads");
+                                    shipping.stop();
+                                    choosing.start();
+                                    ret.clear();
+                                }
+                                currentShardIndex++;
+                                currentShard = shards.get(currentShardIndex);
+                            }
+                            ret.add(r);
+                        }
+                        choosing.stop();
+                        // what remains in ret is sent after the if statement
+                    } else if (cloudStorageUrl) {
+                        // GCS file, load just the bit we need.
+                        querying.start();
+                        final SamReader reader = BAMIO.openBAM(storageClient, bam, ValidationStringency.DEFAULT_STRINGENCY);
+                        SAMRecordIterator query = reader.query(i.getContig(), i.getStart(), i.getEnd(), false);
+                        // we read multiple at once to estimate how long we spend waiting for disk.
+                        BatchingIterator<SAMRecord> batcher = new BatchingIterator<SAMRecord>(query, outputShardSize);
+                        querying.stop();
+
+                        loading.start();
+                        while (batcher.hasNext()) {
+                            ArrayList<SAMRecord> buf = batcher.next();
+                            loading.stop();
+                            choosing.start();
+                            bunny.stepEnd("done loading a batch");
+                            for (SAMRecord r : buf) {
+                                if (!accept(r, i)) {
+                                    rejected++;
+                                    continue;
+                                }
+
+                                while (currentShard.getEnd() < r.getStart()) {
```

Comment:
There's a lot of duplicated code between the local and cloud storage cases -- can you extract a common method?